experimental/ir/fdp.go

```diff
@@ -0,0 +1,443 @@
+	}
+}
+
+func (c *codegen) enum(ty Type, mdp *descriptorpb.EnumDescriptorProto) {
+	mdp.Name = addr(ty.Name()) // Has a leading dot.
+
+	for field := range seq.Values(ty.Fields()) {
+		evd := new(descriptorpb.EnumValueDescriptorProto)
+		mdp.Value = append(mdp.Value, evd)
+		c.enumValue(field, evd)
+	}
+
+	//nolint:revive // Complains about range_.
+	for range_ := range seq.Values(ty.ReservedRanges()) {
+		rr := new(descriptorpb.EnumDescriptorProto_EnumReservedRange)
+		mdp.ReservedRange = append(mdp.ReservedRange, rr)
+
+		start, end := range_.Range()
+		rr.Start = addr(start)
+		rr.End = addr(end - 1)
+	}
+
+	for name := range seq.Values(ty.ReservedNames()) {
+		mdp.ReservedName = append(mdp.ReservedName, name.Name())
+	}
+
+	for option := range seq.Values(ty.Options()) {
+		if mdp.Options == nil {
+			mdp.Options = new(descriptorpb.EnumOptions)
+		}
+
+		c.option(option, mdp.Options)
+	}
+}
+
+func (c *codegen) enumValue(f Field, fdp *descriptorpb.EnumValueDescriptorProto) {
+	fdp.Name = addr(f.Name())
+	fdp.Number = addr(f.Number())
+
+	for option := range seq.Values(f.Options()) {
+		if fdp.Options == nil {
+			fdp.Options = new(descriptorpb.EnumValueOptions)
+		}
+
+		c.option(option, fdp.Options)
+	}
+}
+
+func (c *codegen) option(_ Option, target proto.Message) {
+	var fdp *descriptorpb.FieldDescriptorProto
+	if actual, ok := target.(*descriptorpb.FieldDescriptorProto); ok {
+		fdp = actual
+		target = fdp.Options
+	}
+
+	_ = target
+
+	// There are two cases and both are painful.
+	//
+	// 1. For built-in options, we need to match up option.Field() to a
+	//    protoreflect.Field in target, and then set it.
+	//
+	//    If we recognize this field as a pseudo-option, we need to forgo the
+	//    above and set it directly on the non-nil fdp instead.
+	//
+	// 2. For custom options, we need to serialize option (perhaps with an
+	//    Option.Marshal() function?) and append it to the unknown fields.
+
+	// TODO: Implement the above (ow ow ow).
+}
+
+// addr is a helper for creating a pointer out of any type, because Go is
+// missing the syntax &"foo", etc.
+func addr[T any](v T) *T { return &v }
+
+// Set of all names that are defined in scope of some message; used for
+// generating synthetic names.
+type syntheticNames map[intern.ID]struct{}
+
+func (sn *syntheticNames) generate(candidate string, message Type) string {
+	table := message.Context().intern
+	var names = *sn
+	if *sn == nil {
+		names = make(map[intern.ID]struct{})
+		*sn = names
+
+		for field := range seq.Values(message.Fields()) {
+			names[field.InternedName()] = struct{}{}
+		}
+		for extn := range seq.Values(message.Extensions()) {
+			names[extn.InternedName()] = struct{}{}
+		}
+		for oneof := range seq.Values(message.Oneofs()) {
+			names[oneof.InternedName()] = struct{}{}
+		}
+		for ty := range seq.Values(message.Nested()) {
+			name := ty.Name()
+			name = name[strings.LastIndex(name, ".")+1:]
+			names[table.Intern(name)] = struct{}{}
+		}
```

Comment:
Also need to consider enum values for any nested enums, since enum values follow C++ scoping rules are values are in the same scope as the enum itself (instead of the enum being a new scope, and values being children thereof).